### ISSUE TITLE
Value degradation when calling pngwriter::plot with float values

### ISSUE THREAD

**Issue by mxmlnkn** [pngwriter/pngwriter#83]
I was implementing a unit test for a read/write function using pngwriter and found the following problem. Some input values, like `0.868921399117f` were degrading to smaller and smaller values when chaining calls to `read` and `plot`. (More exactly I really wrote out the png and then read it in anew).

After roughly 120 iterations the value reaches a fix value of `0.868925001907f`. When representing this as the pngwriter internal integer this means the first value of `56945` degraded down to `56831` thereby loosing quite a bit of precision.

The problem is the simple cast to int in `pngwriter::plot(int, int, double, double, double)`. Here is an example reproducing the problem and showing two possible bugfixes.

```C++
    #include <iostream>
    int main(void)
    {
        float originalValue = 0.868921399117;
        for ( int i = 0; i < 120; ++i )
        {
            /* call pngwriter.plot which only exists with double arguments */
            double doubleValueToPlot = double( originalValue );
            /* the double version calls the int version with a cast wich always
             * rounds using floor which results in value degradataion */
            int intToPlot = int( doubleValueToPlot*65535 );
            std::cout << "Plot int: " << intToPlot << "\n";
            /* the int is written out without loss and then normally you would
             * try to reconstruct the original float: */
            float readInValue = (float) intToPlot / 65535;
            originalValue = readInValue;
        }

        std::cout << "=== Bugfix 1 ===\n";
        originalValue = 0.868921399117;
        for ( int i = 0; i < 10; ++i )
        {
            /* add a plot function in pngwriter accepting float */
            float valueToPlot = float( originalValue );
            int intToPlot = int( valueToPlot*65535 );
            std::cout << "Plot int: " << intToPlot << "\n";
            float readInValue = (float) intToPlot / 65535;
            originalValue = readInValue;
        }

        std::cout << "=== Bugfix 2 ===\n";
        originalValue = 0.868921399117;
        for ( int i = 0; i < 10; ++i )
        {
            double doubleValueToPlot = double( originalValue );
            /* round correctly instead of doing a direct floor-like int cast */
            auto round = []( double x ) -> int {
                return (x > 0.0) ? (x + 0.5) : (x - 0.5);
            };
            int intToPlot = round( doubleValueToPlot*65535 );
            std::cout << "Plot int: " << intToPlot << "\n";
            float readInValue = (float) intToPlot / 65535;
            originalValue = readInValue;
        }
    }
```

Note that Bugfix 1 converges to `56944` but bugfix 2 converges to `56945`. The latter is more correct, because the input was: `56944.7638911`

**Comment by ax3l** [comment]
Thanks for the precise report! :sparkles: 

As you correctly found out, the underlying issue is that float to int converts should in principle keep the 5 digits in the  `65535` shades of each color channel ([plot(...)](https://github.com/pngwriter/pngwriter/blob/0.5.6/src/pngwriter.cc#L745-L748)). As an example an expansion/multiplication is applied in the dbl plot member function:

```C++
void pngwriter::plot(int x, int y, double red, double green, double blue)
{
   this->plot(x,y,int(red*65535),int(green*65535),int(blue*65535));
};
```

Currently, pngwriter only writes [16 bit/color channel png images](https://github.com/pngwriter/pngwriter/issues/15) (which is already the highest value on can put into each channel in a png [[1]](https://en.wikipedia.org/wiki/Portable_Network_Graphics#Pixel_format)).

During `read`, did you use the provided [`dread(...)`](https://github.com/pngwriter/pngwriter/blob/0.5.6/src/pngwriter.h#L244-L251) function? It returns floats.

Also be aware that `originalValue = 0.868921399117` is already a number with higher precision as `float32_t` supports. If your value is *very* close to `0.0` even more troubles arise to do to [subnormal number IEEE rounding](https://en.wikipedia.org/wiki/Denormal_number).

Just as a side note to get things done in your unit test: using an *epsilon comparison* or or [float32 tiff's](https://gist.github.com/ax3l/5781ce80b19d7df3f549) or hdf5 might also help.

**Comment by psychocoderHPC** [comment]
I also checked this issue and found that `BUGFIX I` is only working because of the bad float precision.

I think to minimize the rounding error from double to int a round nearest is a valid solution. 
`pngwriter::plot` plot is only valid for doubles from [0;1] therefor a liddle change without performance influence can be added:
```
void pngwriter::plot(int x, int y, double red, double green, double blue)
{
   this->plot(x,y,int(red*65535.5),int(green*65535.5),int(blue*65535.5));
};
```

**Comment by ax3l** [comment]
great, @mxmlnkn if you want to test this we can prepare a PR.

Should we also update the `*256` in the `int read(...)` calls? [[2]](https://github.com/pngwriter/pngwriter/blob/0.5.6/src/pngwriter.cc#L751-L860)

**Comment by psychocoderHPC** [comment]
@ax3l The `read(...)` used only `int` for calculation and need IMO no changes.

**Comment by mxmlnkn** [comment]
> During read, did you use the provided dread(...) function? It returns floats.

No I didn't, but I did now in the unit test and the problem also appears there.

>If your value is very close to 0.0

Well the value closest to 0 not rounding to 0 because of precision would be 1/65535. If that is not small enough, then everything should be fine.

See pull request #84

Btw the output of the unit test before the round to nearest was implemented is:

    [...]
    Input Value  0.348608 converged to 0.347646  in    63 iterations
    Input Value  0.751369 converged to 0.749996  in    90 iterations
    Input Value  0.622278 converged to 0.621088  in    78 iterations
    Input Value  0.205432 converged to 0.205066  in    24 iterations
    Input Value 0.0849622 converged to 0.084947  in     1 iterations
    Input Value  0.527733 converged to 0.527337  in    26 iterations
    Input Value  0.436194 converged to 0.435538  in    43 iterations
    Input Value  0.658183 converged to 0.656245  in   127 iterations
    Input Value  0.957565 converged to 0.957031  in    35 iterations
    Input Value   0.12607 converged to 0.125963  in     7 iterations
    Input Value  0.217899 converged to 0.217762  in     9 iterations
    Input Value  0.614511 converged to 0.613275  in    81 iterations
    Input Value  0.249119 converged to 0.249012  in     7 iterations
    Input Value  0.860868 converged to 0.859373  in    98 iterations
    Input Value  0.566598 converged to   0.5664  in    13 iterations
    Input Value   0.18172 converged to 0.181628  in     6 iterations
    Input Value  0.805447 converged to 0.804685  in    50 iterations
    Input Value  0.841108 converged to 0.839841  in    83 iterations
    Input Value   0.64593 converged to 0.644526  in    92 iterations
    Input Value  0.189731 converged to 0.189441  in    19 iterations
    Input Value  0.758648 converged to 0.757809  in    55 iterations
    Input Value  0.268223 converged to 0.267567  in    43 iterations
    Input Value  0.629023 converged to 0.628901  in     8 iterations
    Input Value  0.676417 converged to 0.675776  in    42 iterations
    Input Value  0.488655 converged to 0.488273  in    25 iterations
    Input Value  0.379324 converged to 0.378897  in    28 iterations
    Input Value  0.307012 converged to  0.30663  in    25 iterations
    Input Value  0.578225 converged to 0.578119  in     7 iterations
    Input Value  0.094789 converged to 0.0947127 in     5 iterations
    Input Value  0.433677 converged to 0.433585  in     6 iterations
    Input Value  0.192737 converged to  0.19237  in    24 iterations
    Found 497 problematic values out of 1000

I find this quite interesting, because 497 out of 1000 would mean that the problem is very likely only 1 bit.

**Comment by ax3l** [comment]
fixed in #84
